### PR TITLE
Add Hierarchical Clustering with Linkage Options

### PR DESCRIPTION
**Pull request by Janne98**
This update introduces hierarchical clustering. Users can specify the linkage method (`--linkage`), with options:

- single
- average
- complete (default)

Density based clustering (DBSCAN) was removed, because the results are almost identical to single linkage hierarchical clustering.

### CODE REVIEW

File: falcon/cluster/cluster.py [hunk 1]
```diff
@@ -920,3 +746,76 @@ def _get_cluster_medoid_index(
...
+    """
+    Compute the condensed pairwise distance matrix for the given spectra.
+
+    Parameters
+    ----------
+    spec_tuples : List[similarity.SpectrumTuple]
+        The spectra to compute the pairwise distance matrix for.
+    fragment_mz_tolerance : float
+        The fragment m/z tolerance.
+    allow_shift : bool
```
Comment (bittremieux) [inline falcon/cluster/cluster.py:765]: **todo:** This can be removed. During clustering we're only considering near-identical spectra from the same molecule, no shifted peaks.

File: falcon/cluster/cluster.py [hunk 2]
```diff
@@ -920,3 +746,76 @@ def _get_cluster_medoid_index(
...
+    np.ndarray
+        The condensed pairwise distance matrix.
+    """
+    n = len(spec_tuples)
+    condensed_dist_matrix = np.zeros(n * (n - 1) // 2)
+
+    def worker(i, j):
+        spec_tup1 = spec_tuples[i]
+        spec_tup2 = spec_tuples[j]
+        sim, _, _, _, _, _, _ = similarity._cosine_fast(
```
Comment (bittremieux) [inline falcon/cluster/cluster.py:779]: **todo:** Simplify return values to only what's actually necessary.

File: falcon/cluster/cluster.py [hunk 3]
```diff
@@ -920,3 +746,76 @@ def _get_cluster_medoid_index(
...
+            spec_tup1, spec_tup2, fragment_mz_tol, allow_shift
+        )
+        distance = 1.0 - sim
+        idx = condensed_index(i, j, n)
+        condensed_dist_matrix[idx] = abs(distance)
```
Comment (bittremieux) [inline falcon/cluster/cluster.py:784]: **question:** Not sure I understand this. Why is this necessary? The distance should always be >= 0, so what is this supposed to do?

File: falcon/cluster/similarity.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+)
+
+#  return type
+SimilarityTuple = collections.namedtuple(
+    "SimilarityTuple",
+    [
+        "score",
+        "matched_intensity",
+        "max_contribution",
+        "n_greq_2p",  # signals contributing >= 2% score
```
Comment (bittremieux) [inline falcon/cluster/similarity.py:22]: **todo:** Remove stuff that's not necessary.

File: falcon/cluster/similarity.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+    Returns
+    -------
+    SimilarityTuple
+        A tuple consisting of the cosine similarity between both spectra,
+        matched intensity, maximum contribution by a signal pair, matched
+        signals, and arrays of the matching peak indexes in the first and
+        second spectrum.
+    """
+    spec_tup1 = _df_row_to_spectrum_tuple(spectrum1)
```
Comment (bittremieux) [inline falcon/cluster/similarity.py:58]: **question:** Can we convert directly to spectrum tuples instead of having to make the roundtrip via a Series?

File: falcon/cluster/similarity.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    Parameters
+    ----------
+    spec : SpectrumTuple
+        Numba-compatible tuple containing information from the first spectrum.
+    spec_other : SpectrumTuple
+        Numba-compatible tuple containing information from the second spectrum.
+    fragment_mz_tolerance : float
+        The fragment m/z tolerance used to match peaks in both spectra with
+        each other.
+    allow_shift : bool
```
Comment (bittremieux) [inline falcon/cluster/similarity.py:84]: **todo:** Simplify to not allow peak shifts. Might make the implementation a bit (probably not a lot) faster as well.

**Review by bittremieux — CHANGES_REQUESTED** [review]

File: falcon/cluster/cluster.py (hunk 3 above)
Comment (Janne98) [reply falcon/cluster/cluster.py:784]: Previously, I ran into the issue of the distance being slightly less than 0 for some reason, but that doesn't seem to be the case anymore since using `_cosine_fast`

File: falcon/cluster/similarity.py (hunk 5 above)
Comment (Janne98) [reply falcon/cluster/similarity.py:58]: The `cosine` function isn't even used anymore. I directly call `_cosine_fast`

File: falcon/cluster/cluster.py [hunk 7]
```diff
@@ -1,623 +1,395 @@
...
-            ],
-            scan_in_order=True,
+                "mz",
+                "intensity",
+            ]
         )
         .to_pandas()
-        .rename(columns={"identifier": "spectrum_id"})
+        .reset_index()
+        .sort_values("precursor_mz")
```
[the remark this answers is not in the document]
Comment (Janne98) [reply falcon/cluster/cluster.py:139]: I can sort the pyarrow table, but not the lance dataset

File: falcon/cluster/cluster.py (hunk 3 above)
Comment (bittremieux) [reply falcon/cluster/cluster.py]: Ok. For future reference, using a "clip" to 0 behavior is better in that case. The absolute value will be non-zero positive, but presumably those negative values are due to some rounding errors and should be just 0 instead.

Comment (Janne98) [reply falcon/cluster/cluster.py]: It's still a problem on our toy dataset. Some similarities are just slightly > 1. I'll clip in `_cosine_fast`.

File: falcon/cluster/cluster.py [hunk 8]
```diff
@@ -2,622 +2,341 @@
...
+    # precursor m/z threshold).
+    logging.info(
+        "Cluster %d spectra using %s linkage and distance threshold %.3f",
+        len(data),
+        linkage,
+        distance_threshold,
+    )
+    with tempfile.NamedTemporaryFile(suffix=".npy") as cluster_file:
+        cluster_filename = cluster_file.name
+    cluster_labels = np.lib.format.open_memmap(
```
Comment (bittremieux) [inline falcon/cluster/cluster.py:94]: **todo:** I think this code should be within the `with` block, because after you exit it, the temporary file is not guaranteed to keep existing but can be cleaned up by the OS. Alternatively, don't use a context manager.

File: falcon/cluster/cluster.py [hunk 9]
```diff
@@ -2,622 +2,341 @@
...
                 "retention_time",
-            ],
-            scan_in_order=True,
+                "mz",
+                "intensity",
+            ]
         )
+        .add_column(0, "index", pa.array(range(dataset.count_rows())))
+        .sort_by("precursor_mz")
         .to_pandas()
```
Comment (bittremieux) [inline falcon/cluster/cluster.py:82]: **todo:** Actually maybe converting to Pandas first is better. I assume it's a view on the data anyway. Because now you're explicitly adding an `index` column, but Pandas will always also create an index. By sorting the Pandas dataframe you can just use its index and avoid adding an entire column.

File: falcon/cluster/similarity.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    Returns
+    -------
+    SpectrumTuple
+        A tuple containing the spectrum information.
+    """
+    return SpectrumTuple(
+        row["precursor_mz"],
+        row["precursor_charge"],
+        row["mz"],
+        np.copy(row["intensity"]) / np.linalg.norm(row["intensity"]),
```
Comment (bittremieux) [inline falcon/cluster/similarity.py:101]: **question:** This is normalizing the intensity every time a spectrum is used to calculate the cosine, potentially doing this multiple times for the same spectrum. Is this not being done during preprocessing yet? If not, can it be done there?

**Review by bittremieux — CHANGES_REQUESTED** [review]

File: falcon/cluster/similarity.py (hunk 10 above)
Comment (Janne98) [reply falcon/cluster/similarity.py:101]: This is indeed already being done during preprocessing

File: falcon/cluster/cluster.py [hunk 11]
```diff
@@ -2,622 +2,343 @@
...
                 "retention_time",
-            ],
-            scan_in_order=True,
+                "mz",
+                "intensity",
+            ]
         )
         .to_pandas()
-        .rename(columns={"identifier": "spectrum_id"})
+        .reset_index()
```
Comment (bittremieux) [inline falcon/cluster/cluster.py:81]: **todo:** You don't actually need to reset the index (duplicating the information into an "index" column and the new index), but can use the index directly instead. So line 102 would become `idx = data.index`.

**Review by bittremieux — APPROVED** [review]
Looks good. One more minor suggestion and then it's ready to be merged.
